src/jdk.jartool/share/classes/sun/security/tools/jarsigner/Main.java

```diff
@@ -1128,6 +1133,138 @@ void verifyJar(String jarName)
         }
     }
 
+    private void crossCheckEntries(String jarName) throws Exception {
+        List<String> locEntries = new ArrayList<>();
+        List<String> cenEntries;
+
+        try (JarFile jarFile = new JarFile(jarName);
+             JarInputStream jis = new JarInputStream(
+                     Files.newInputStream(Path.of(jarName)))) {
+
+            Manifest cenManifest = jarFile.getManifest();
+            Manifest locManifest = jis.getManifest();
+            if (!compareManifest(cenManifest, locManifest)) {
+                return;
+            }
+
+            JarEntry locEntry;
+            while ((locEntry = jis.getNextJarEntry()) != null) {
+                String entryName = locEntry.getName();
+                locEntries.add(entryName);
+
+                JarEntry cenEntry = jarFile.getJarEntry(entryName);
+                if (cenEntry == null) {
+                    crossChkWarnings.add(String.format(rb.getString(
+                            "Entry.1.present.when.reading.jarinputstream." +
+                            "but.missing.via.JarFile"), entryName));
+                    continue;
+                }
+
+                readEntry(jis);
+                try (InputStream cenInputStream = jarFile.getInputStream(cenEntry)) {
+                    if (cenInputStream == null) {
+                        crossChkWarnings.add(String.format(rb.getString(
+                                "entry.1.present.in.jarfile.but.unreadable"),
+                                entryName));
+                    } else {
+                        readEntry(cenInputStream);
```

Comment:
This could throw a `SecurityException` if the content does not match the digest. Although we have verified the JAR file with`JarFile`, things could be different when opened with a `JarInputStream`.

Reply:
Added try-catch block.